subsys/bluetooth/mesh/sensor_types.c

```diff
@@ -471,6 +514,244 @@ static int float32_decode(const struct bt_mesh_sensor_format *format,
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+static bool percentage_delta_check(const struct bt_mesh_sensor_value *delta,
+				   float diff, float prev)
+{
+	if (prev == 0.0f) {
+		return true;
+	}
+
+	float percentage_float = scalar_decode_raw(delta->format, delta->raw) / 10000.0f;
+
+	return (diff / prev) > (percentage_float);
+}
+
+static bool scalar_delta_check(const struct bt_mesh_sensor_value *current,
+			       const struct bt_mesh_sensor_value *previous,
+			       const struct bt_mesh_sensor_deltas *deltas)
+{
+	const struct bt_mesh_sensor_value *delta;
+
+	int64_t current_raw = scalar_decode_raw(current->format, current->raw);
+	int64_t previous_raw = scalar_decode_raw(previous->format, previous->raw);
+	int64_t diff = current_raw - previous_raw;
+
+	if (diff < 0) {
+		diff = -diff;
+		delta = &deltas->down;
+	} else {
+		delta = &deltas->up;
+	}
+
+	int64_t delta_raw = scalar_decode_raw(delta->format, delta->raw);
+
+	if (delta->format == &bt_mesh_sensor_format_percentage_delta_trigger) {
+		return percentage_delta_check(delta, diff, previous_raw);
+	}
+
+	return diff > delta_raw;
+}
+
+static int boolean_compare(const struct bt_mesh_sensor_value *op1,
+			   const struct bt_mesh_sensor_value *op2)
+{
+	int16_t diff = *(op1->raw) - *(op2->raw);
+
+	if (diff > 0) {
+		return 1;
+	} else if (diff < 0) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+static bool boolean_delta_check(const struct bt_mesh_sensor_value *current,
+				const struct bt_mesh_sensor_value *previous,
+				const struct bt_mesh_sensor_deltas *deltas)
+{
+	uint8_t current_raw =  *(current->raw);
+	uint8_t previous_raw = *(previous->raw);
+	uint8_t diff;
+	const struct bt_mesh_sensor_value *delta;
+
+	if (current_raw < previous_raw) {
+		diff = previous_raw - current_raw;
+		delta = &deltas->down;
+	} else {
+		diff = current_raw - previous_raw;
+		delta = &deltas->up;
+	}
+
+	if (delta->format == &bt_mesh_sensor_format_percentage_delta_trigger) {
+		return percentage_delta_check(delta, diff, previous_raw);
+	}
+
+	return diff > *(delta->raw);
+}
+
+static int exp_1_1_compare(const struct bt_mesh_sensor_value *op1,
+			   const struct bt_mesh_sensor_value *op2)
+{
+	uint8_t op1_raw = *(op1->raw);
+	uint8_t op2_raw = *(op2->raw);
+
+	if (op1_raw > op2_raw) {
+		return 1;
+	} else if (op1_raw < op2_raw) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+/* This is the negative of the ceiling of the function d_1.1(-(i+1)), where d_1.1
+ * is the Gaussian logarithm difference function d_b(z) = log_b(|1 - b^z|).
+ * -ceil(d_1.1(-(i+1))) is 0 for i+1 > 25.
+ */
+static uint8_t gaussian_log_d11[] = { 25, 18, 14, 12, 10, 8, 7, 6, 5, 5, 4, 4,
+				      3, 3, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1 };
+
+/** Computes ceil(log_1.1(|1.1^(a-64) - 1.1^(b-64)|)) + 64 using Gaussian logarithms.
+ *  a must be greater than b.
+ */
+static uint8_t exp_1_1_ceil_diff(uint8_t a, uint8_t b)
+{
+	uint8_t diff = a - b;
+	uint8_t d = diff > 25 ? 0 : gaussian_log_d11[diff - 1];
```

Comment:
_I gonna tell the compiler about the comment..._ Please add an assertion at least.

Reply:
Then it's easier to just `if (diff == 0) { return 0; }` tbh, makes it more general and avoids risk of out-of-bounds array access.

Reply:
If it is assumed that `a` must be greater than `b`, then returning 0 will simply ignore the actual error.

Reply:
No, 0 is the correct answer if a == b. raw = 0 is defined in the spec to represent 0.0 (and not 1.1^-64). I just added the assumption to save three lines of code since it's only called from one function where the a == b check is already done before the call is reached.

Reply:
Ok